Userland/Libraries/LibHID/ReportDescriptorParser.h

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2025, Sönke Holz <sholz8530@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Endian.h>
+#include <AK/Forward.h>
+#include <AK/HashMap.h>
+#include <AK/MemoryStream.h>
+#include <AK/SetOnce.h>
+#include <AK/Stack.h>
+
+#include <LibHID/ReportDescriptorDefinitions.h>
+
+namespace HID {
+
+// https://www.usb.org/document-library/device-class-definition-hid-111
+
+class ItemStream : public FixedMemoryStream {
+public:
+    using FixedMemoryStream::FixedMemoryStream;
+
+    ErrorOr<ItemHeader> read_item_header()
+    {
+        return read_value<ItemHeader>();
+    }
+
+    template<typename T>
+    ErrorOr<T> read_item_data(ItemHeader item_header)
+    {
+        VERIFY(item_header.type != ItemType::Reserved && item_header.tag != TAG_LONG_ITEM);
+
+        if (item_header.real_size() > sizeof(T))
+            return Error::from_errno(EINVAL);
+
+        // Short items have a max data length of 4.
+        alignas(T) u8 buffer[4] {};
+        TRY(read_until_filled({ &buffer, item_header.real_size() }));
```

Comment:
(this assumes little endian machines, right? which is fine, but the int readers below are explicit about it)

Reply:
This function doesn't assume any endianness. It just reads raw binary data into a buffer.
If you want to read a struct containing some multibyte int value you would have to wrap it in a `*Endian<T>`.

Reply:
The HID class definition actually doesn't specify the endianness of report descriptor items, only the endianness of the report fields. But it seems to be little endian.